apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -63,187 +64,459 @@ defmodule Champions.Battle.Simulator do
+  defp check_winner(state, step) do
+    winner =
+      cond do
+        Enum.empty?(state.units) -> :tie
+        Enum.all?(state.units, fn {_id, unit} -> unit.team == 2 end) -> :team_2
+        Enum.all?(state.units, fn {_id, unit} -> unit.team == 1 end) -> :team_1
+        true -> :none
+      end
+
+    case winner do
+      :none ->
+        if step == @maximum_steps do
+          Logger.info("Battle timeout.")
+          {:halt, :timeout}
+        else
+          {:cont, state}
+        end
+
+      result ->
+        Logger.info("Battle ended. Result: #{result}.")
+        {:halt, result}
+    end
+  end
+
+  # Calculate the new state of the battle after a step passes for a unit.
+  # Updates cooldowns, casts skills and reduces self-affecting modifier durations.
+  defp process_step_for_unit(unit, current_state, initial_step_state) do
     new_state =
       cond do
-        not can_attack(unit) ->
+        not can_attack(unit, initial_step_state) ->
+          Logger.info("Unit #{format_unit_name(unit)} cannot attack")
           current_state
 
         can_cast_ultimate_skill(unit) ->
-          unit
-          |> cast_skill(unit.ultimate_skill, initial_step_state, current_state)
-          |> put_in([unit.id, :energy], 0)
+          Logger.info("Unit #{format_unit_name(unit)} casting Ultimate skill")
+
+          current_state
+          |> Map.put(:skills_being_cast, [unit.ultimate_skill | current_state.skills_being_cast])
+          |> put_in([:units, unit.id, :energy], 0)
 
         can_cast_basic_skill(unit) ->
-          new_state = cast_skill(unit, unit.basic_skill, initial_step_state, current_state)
+          Logger.info("Unit #{format_unit_name(unit)} casting basic skill")
 
-          new_state
+          current_state
+          |> Map.put(:skills_being_cast, [unit.basic_skill | current_state.skills_being_cast])
           |> put_in(
-            [unit.id, :basic_skill, :remaining_cooldown],
+            [:units, unit.id, :basic_skill, :remaining_cooldown],
             unit.basic_skill.base_cooldown + 1
           )
-          |> put_in([unit.id, :energy], new_state[unit.id][:energy] + 75)
+          |> put_in([:units, unit.id, :energy], current_state.units[unit.id].energy + unit.basic_skill.energy_regen)
 
         true ->
           current_state
       end
 
-    put_in(
-      new_state,
-      [unit.id, :basic_skill, :remaining_cooldown],
-      max(new_state[unit.id].basic_skill.remaining_cooldown - 1, 0)
-    )
+    # Reduce modifier remaining timers & remove expired ones
+    new_modifiers = %{
+      additives: reduce_modifier_timers(unit.modifiers.additives, unit),
+      multiplicatives: reduce_modifier_timers(unit.modifiers.multiplicatives, unit),
+      overrides: reduce_modifier_timers(unit.modifiers.overrides, unit)
+    }
 
-    # TODO: decrease_effects_remaining_steps(unit, current_state)
+    # Reduce basic skill cooldown
```

Comment:
If you handle cooldowns first you could remove the +1 that this put has `remove_dead_units(new_state)
      |> check_winner(step)`

Reply:
I'd rather not sacrifice any more performance than necessary since battles could be susceptible to turning too long to process, so I think going through the units list once will be better than handling it separately. On handling the cooldowns before processing, I think that could work! I'll work on it

Reply:
We talked about cooldown and we couldn't find a better solution to the +1 issue, so we'll just document it and go on.

Also i don't think we should optimize before hand, maybe reaching a first iteration of the game is more important than make it performant.